Orange/widgets/utils/owlearnerwidget.py

```diff
@@ -1,26 +1,207 @@
+        self.data = None
+        self.learner = None
+        self.model = None
+        self.preprocessors = None
+        self.outdated_settings = False
+        self.setup_layout()
+        self.apply()
+
+    def set_preprocessor(self, preprocessor):
         """Add user-set preprocessors before the default, mandatory ones"""
-        self.preprocessors = ((preproc,) if preproc else ()) + tuple(self.LEARNER.preprocessors)
+        self.preprocessors = ((preprocessor,) if preprocessor else ()) + tuple(self.LEARNER.preprocessors)
         self.apply()
 
+    @check_sql_input
+    def set_data(self, data):
+        """Set the input train data set."""
+        self.error(self.DATA_ERROR_ID)
+        self.data = data
+        if data is not None and data.domain.class_var is None:
+            self.error(self.DATA_ERROR_ID, "Data has no target variable")
+            self.data = None
 
-class OWProvidesLearner(_OWAcceptsPreprocessor):
-    """
-    Base class for all classification / regression learner-providing widgets
-    that extend it.
-    """
+        self.update_model()
+
+    def apply(self):
+        """Applies leaner and sends new model.
+        """
+        self.update_learner()
+        self.update_model()
+
+    @abstract
+    def create_learner(self):
+        """Creates a learner with current configuration.
+
+        Returns:
+            Leaner: an instance of Orange.base.learner subclass.
+        """
+        raise NotImplementedError()
+
+    def update_learner(self):
+        self.learner = self.create_learner()
+        self.send("Learner", self.learner)
+        self.outdated_settings = False
+        self.warning(self.OUTDATED_LEARNER_WARNING_ID)
+
+    def update_model(self):
+        self.model = None
+        self.good_data = False
+        if self.data is not None:
+            self.error(self.DATA_ERROR_ID)
+            if not self.learner.check_learner_adequacy(self.data.domain):
+                self.error(self.DATA_ERROR_ID, self.learner.learner_adequacy_err_msg)
+            elif len(np.unique(self.data.Y)) < 2:
+                self.error(self.DATA_ERROR_ID,
+                           "Data contains single target value. "
+                           "There is nothing to learn.")
+            else:
+                self.model = self.learner(self.data)
+                self.model.name = self.learner_name
+                self.model.instances = self.data
+                self.good_data = True
+
+        self.send(self.OUTPUT_MODEL_NAME, self.model)
+
+    def settings_changed(self, *args, **kwargs):
+        self.outdated_settings = True
+        self.warning(self.OUTDATED_LEARNER_WARNING_ID,
+                     "Press Apply to submit changes.")
+
+    def get_model_parameters(self):
+        """Creates an `OrderedDict` or a sequence of pairs with current model
+        configuration.
+
+        Returns:
+            OrderedDict or List: (option, value) pairs or dict
+        """
+        return None
```

Comment:
Can also return `None`? I guess let's keep the code more invariant by returning an empty list-like iterable by default, e.g. `[]`.